File: src/main/kotlin/com/rainbow/cache/druid/DruidDataSourceConfig.java
package com.rainbow.cache.druid;

import com.alibaba.druid.pool.DruidDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.bind.RelaxedPropertyResolver;
import org.springframework.context.EnvironmentAware;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.env.Environment;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import javax.sql.DataSource;
import java.sql.SQLException;

/**
 * Created by Administrator on 2017/6/17.
 */
@Configuration
@EnableTransactionManagement
public class DruidDataSourceConfig implements EnvironmentAware {

    private static final Logger logger = LoggerFactory.getLogger(DruidDataSourceConfig.class);

    private RelaxedPropertyResolver propertyResolver;

    @Override
    public void setEnvironment(Environment environment) {
        this.propertyResolver = new RelaxedPropertyResolver(environment, "spring.datasource");
    }

    @Bean
    public DataSource dataSource() {
        logger.info("开始注入druid");
        DruidDataSource dataSource = new DruidDataSource();

        dataSource.setValidationQuery(propertyResolver.getProperty("validationQuery"));
        dataSource.setPassword(propertyResolver.getProperty("password"));
        dataSource.setUsername(propertyResolver.getProperty("username"));
        dataSource.setUrl(propertyResolver.getProperty("url"));
        dataSource.setDriverClassName(propertyResolver.getProperty("driver-class-name"));
        try {
            dataSource.setFilters("stat,wall");
        } catch (SQLException e) {
            logger.error(e.getMessage());
        }
        return dataSource;
    }
}

File: src/main/kotlin/com/rainbow/cache/druid/DruidDataSourceConfig.kt
package com.rainbow.cache.druid

import com.alibaba.druid.pool.DruidDataSource
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.boot.web.servlet.ServletComponentScan
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import java.sql.SQLException
import javax.sql.DataSource

/**
 * Created by rainbow on 2017/7/3.
 *一事专注，便是动人；一生坚守，便是深邃！
 */
@Configuration
@ConditionalOnClass(com.alibaba.druid.pool.DruidDataSource::class)
@ConditionalOnProperty(name = arrayOf("spring.datasource.type"), havingValue = "com.alibaba.druid.pool.DruidDataSource", matchIfMissing = true)
@ServletComponentScan("com.rainbow.filter")
class DruidDataSourceConfig {

    private val logger by lazy { LoggerFactory.getLogger(DruidDataSourceConfig::class.java) }


    @Autowired
    lateinit private var druidDataSourceProperty: DruidDataSourceProperty

    @Bean
    @Primary
    fun dataSource(): DataSource {


        logger.info("开始初始化druid数据源")

        val datasource = DruidDataSource()

        datasource.url = druidDataSourceProperty.url
        datasource.password = druidDataSourceProperty.password
        datasource.driverClassName = druidDataSourceProperty.driverClassName
        datasource.username = druidDataSourceProperty.username
        datasource.initialSize = druidDataSourceProperty.initialSize
        datasource.minIdle = druidDataSourceProperty.minIdle
        datasource.maxActive = druidDataSourceProperty.maxActive
        datasource.maxWait = druidDataSourceProperty.maxWait
        datasource.timeBetweenEvictionRunsMillis = druidDataSourceProperty.timeBetweenEvictionRunsMillis
        datasource.minEvictableIdleTimeMillis = druidDataSourceProperty.minEvictableIdleTimeMillis
        datasource.validationQuery = druidDataSourceProperty.validationQuery
        datasource.isTestWhileIdle = druidDataSourceProperty.testWhileIdle
        datasource.isTestOnBorrow = druidDataSourceProperty.testOnBorrow
        datasource.isTestOnReturn = druidDataSourceProperty.testOnReturn
        datasource.isPoolPreparedStatements = druidDataSourceProperty.poolPreparedStatements
        datasource.maxPoolPreparedStatementPerConnectionSize = druidDataSourceProperty.maxPoolPreparedStatementPerConnectionSize
        datasource.isUseGlobalDataSourceStat = druidDataSourceProperty.useGlobalDataSourceStat

        try {
            datasource.setFilters(druidDataSourceProperty.filters)
        } catch (e: SQLException) {
            logger.error("druid configuration initialization filter", e)
        }
        datasource.connectProperties = druidDataSourceProperty.connectionProperties


        return datasource

    }
}